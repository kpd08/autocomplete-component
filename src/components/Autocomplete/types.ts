enum Keys {
  ArrowDown = "ArrowDown",
  ArrowUp = "ArrowUp",
  Enter = "Enter",
}

enum EntryType {
  Repo = "repo",
  User = "user",
}

interface GithubUser {
  login: string;
  html_url: string;
}

interface GithubRepo {
  name: string;
  html_url: string;
}

interface AutcompleteEntry {
  name: string;
  type: EntryType;
  url: string;
}

interface GithubSearchResponse<T = any> {
  total_count: number;
  items: Array<T>;
  incomplete_results: boolean;
}

export type { AutcompleteEntry, GithubUser, GithubSearchResponse, GithubRepo };
export { Keys, EntryType };
