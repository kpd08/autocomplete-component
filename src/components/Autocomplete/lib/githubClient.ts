import { GithubRepo, GithubSearchResponse, GithubUser } from "../types";

const API_URL = process.env.REACT_APP_GITHUB_SEARCH_API;

const defaultOptions: RequestInit = {
  headers: {
    Accept: "application/vnd.github+json",
  },
};

const defaultParams = { per_page: "50" };

const createGetRequest =
  <T = any>(path: string) =>
  async (phrase: string, params = defaultParams): Promise<Array<T>> => {
    try {
      const query = new URLSearchParams({
        ...params,
        q: encodeURIComponent(phrase),
      });
      const url = `${API_URL}${path}?${query.toString()}`;
      const res = await fetch(url, defaultOptions);
      if (res.ok) {
        const { items }: GithubSearchResponse<T> = await res.json();
        return items;
      } else {
        throw new Error();
      }
    } catch (error) {
      throw new Error("Request has failed");
    }
  };

interface GithubSearchClient {
  fetchRepos: (query: string) => Promise<Array<GithubRepo>>;
  fetchUsers: (query: string) => Promise<Array<GithubUser>>;
}

const githubSearchClient: GithubSearchClient = {
  fetchRepos: createGetRequest<GithubRepo>("/repositories"),
  fetchUsers: createGetRequest<GithubUser>("/users"),
};

export default githubSearchClient;
