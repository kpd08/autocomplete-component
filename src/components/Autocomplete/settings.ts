const MIN_PHRASE_LENGTH = 3;
const MAX_ENTRIES_COUNT = 50;
const DEBOUNCE_WAIT_MS = 300;

export { MIN_PHRASE_LENGTH, MAX_ENTRIES_COUNT, DEBOUNCE_WAIT_MS };
