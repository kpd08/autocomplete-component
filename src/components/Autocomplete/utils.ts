import { MAX_ENTRIES_COUNT } from "./settings";
import { GithubRepo, GithubUser, AutcompleteEntry, EntryType } from "./types";

const mapUsersAndReposToResults = (
  users: Array<GithubUser> = [],
  repos: Array<GithubRepo> = []
): Array<AutcompleteEntry> => {
  const usersResults: Array<AutcompleteEntry> = users.map((user) => ({
    name: user.login,
    type: EntryType.User,
    url: user.html_url,
  }));

  const reposResults: Array<AutcompleteEntry> = repos.map((repo) => ({
    name: repo.name,
    type: EntryType.Repo,
    url: repo.html_url,
  }));

  return usersResults
    .concat(reposResults)
    .sort((a, b) => a.name.localeCompare(b.name))
    .slice(0, MAX_ENTRIES_COUNT);
};

export { mapUsersAndReposToResults };
