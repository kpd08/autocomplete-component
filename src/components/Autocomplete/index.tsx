import {
  ChangeEvent,
  FC,
  useState,
  KeyboardEvent,
  useRef,
  useEffect,
  useMemo,
} from "react";
import classNames from "classnames";
import { Keys, AutcompleteEntry, EntryType } from "./types";
import { DEBOUNCE_WAIT_MS, MIN_PHRASE_LENGTH } from "./settings";
import { mapUsersAndReposToResults } from "./utils";
import githubSearchClient from "./lib/githubClient";
import debounce from "lodash.debounce";

interface AutocompleteProps {}

const Autocomplete: FC<AutocompleteProps> = () => {
  const [inputValue, setInputValue] = useState("");
  const [isLoading, setIsLoading] = useState(false);
  const [errorMessage, setErrorMessage] = useState<string | null>(null);
  const [entries, setEntries] = useState<Array<AutcompleteEntry>>([]);
  const [activeEntry, setActiveEntry] = useState<AutcompleteEntry | null>(null);
  const [emptyMessage, setEmptyMessage] = useState<string | null>(null);
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    inputRef.current?.focus();
  }, []);

  useEffect(() => {
    if (activeEntry) {
      const activeElement = document.querySelector(
        "[data-autocomplete-active=true]"
      );
      (activeElement as HTMLAnchorElement)?.scrollIntoView({
        block: "nearest",
        behavior: "auto",
      });
    }
  }, [activeEntry]);

  const handleInputChange = async (evt: ChangeEvent<HTMLInputElement>) => {
    const phrase = evt.target.value;
    setInputValue(phrase);
    if (phrase.length < MIN_PHRASE_LENGTH) {
      clearState(true);
    } else {
      fetchData(phrase);
    }
  };

  const fetchData = useMemo(
    () =>
      debounce(async (phrase: string) => {
        try {
          setIsLoading(true);
          clearState();
          const [users, repos] = await Promise.all([
            githubSearchClient.fetchUsers(phrase),
            githubSearchClient.fetchRepos(phrase),
          ]);
          const results = mapUsersAndReposToResults(users, repos);
          const isEmpty = results.length === 0;
          setEmptyMessage(isEmpty ? "Repositories or users not found" : null);
          setActiveEntry(isEmpty ? null : results[0]);
          setEntries(results);
        } catch (error) {
          const defaultErrorMessage = "Something went wrong";
          if (error instanceof Error) {
            setErrorMessage(error?.message || defaultErrorMessage);
          } else {
            setErrorMessage(defaultErrorMessage);
          }
        } finally {
          setIsLoading(false);
        }
      }, DEBOUNCE_WAIT_MS),
    []
  );

  const handleInputKeyDown = (evt: KeyboardEvent<HTMLInputElement>) => {
    if (evt.key === Keys.Enter && activeEntry) {
      evt.preventDefault();
      window.open(activeEntry.url, "_blank");
    }

    if (evt.key === Keys.ArrowDown && entries.length) {
      evt.preventDefault();
      const index = entries.findIndex((result) => result === activeEntry);
      const nextItem = entries[index + 1] || entries[0];
      setActiveEntry(nextItem);
    }

    if (evt.key === Keys.ArrowUp && entries.length) {
      evt.preventDefault();
      const index = entries.findIndex((result) => result === activeEntry);
      const nextItem = entries[index - 1] || entries[entries.length - 1];
      setActiveEntry(nextItem);
    }
  };

  const clearState = (clearEntries?: boolean) => {
    setEmptyMessage(null);
    setActiveEntry(null);
    setErrorMessage(null);
    if (clearEntries) {
      setEntries([]);
    }
  };

  const renderResults = () => {
    if (emptyMessage) {
      return <div className="text-center p-3">{emptyMessage}</div>;
    }

    return entries.map((entry) => (
      <a
        className={classNames(
          "flex flex-nowrap justify-between items-center w-full p-3 outline-none",
          { "bg-slate-50": activeEntry === entry }
        )}
        key={entry.url}
        href={entry.url}
        target="_blank"
        rel="noreferrer"
        data-autocomplete-active={activeEntry === entry}
        onMouseEnter={() => setActiveEntry(entry)}
      >
        <span className="overflow-clip">{entry.name}</span>
        <span
          className={classNames(
            "ml-2 rounded px-2 py-1 text-xs flex items-center capitalize text-white",
            {
              "bg-green-600": entry.type === EntryType.User,
              "bg-blue-600": entry.type === EntryType.Repo,
            }
          )}
        >
          {entry.type}
        </span>
      </a>
    ));
  };

  return (
    <div className="flex flex-col items-center">
      <input
        onChange={handleInputChange}
        value={inputValue}
        placeholder="Search usernames or repositories"
        className={classNames(
          "border-2 rounded p-4 max-w-md w-full outline-none",
          {
            "border-red-500": errorMessage,
            "rounded-b-none": isLoading || errorMessage || entries.length,
          }
        )}
        ref={inputRef}
        onKeyDown={handleInputKeyDown}
      />
      <div className="flex flex-col max-w-md w-full max-h-autocompleteResults overflow-auto shadow-md rounded-b-md">
        {isLoading && <div className="text-center p-3">loading...</div>}
        {!isLoading && errorMessage && (
          <div className="text-center p-3">{errorMessage}</div>
        )}
        {!isLoading && !errorMessage && renderResults()}
      </div>
    </div>
  );
};

export default Autocomplete;
