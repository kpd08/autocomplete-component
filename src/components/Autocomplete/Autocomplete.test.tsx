import Autocomplete from ".";
import { GithubRepo, GithubUser } from "./types";
import { render, screen } from "@testing-library/react";
import githubClient from "./lib/githubClient";
import userEvent from "@testing-library/user-event";

const mockRepos: Array<GithubRepo> = [
  { name: "a_repo", html_url: "a_repo.url" },
  { name: "b_repo", html_url: "b_repo.url" },
];

const mockUsers: Array<GithubUser> = [
  { login: "a_user", html_url: "a_user.url" },
  { login: "b_user", html_url: "b_user.url" },
];

const fetchUsersSpy = jest.spyOn(githubClient, "fetchUsers");
const fetchReposSpy = jest.spyOn(githubClient, "fetchRepos");
const querySelectorSpy = jest.spyOn(document, "querySelector");

describe("Autocomplete", () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render as expected", () => {
    expect(render(<Autocomplete />).container).toMatchSnapshot();
  });

  it("should display results", async () => {
    jest.useFakeTimers();
    const scrollMock = jest.fn();
    fetchReposSpy.mockResolvedValue(mockRepos);
    fetchUsersSpy.mockResolvedValue(mockUsers);
    querySelectorSpy.mockReturnValue({ scrollIntoView: scrollMock } as any);

    render(<Autocomplete />);

    const phrase = "abc";
    const input = screen.getByPlaceholderText(
      "Search usernames or repositories"
    );
    await userEvent.type(input, phrase);
    await screen.findByText("loading...");
    const links = await screen.findAllByRole("link");

    expect(fetchReposSpy).toHaveBeenNthCalledWith(1, phrase);
    expect(fetchUsersSpy).toHaveBeenNthCalledWith(1, phrase);
    expect(links[0]).toHaveAttribute("data-autocomplete-active", "true");
    expect(links[0]).toHaveAttribute("href", "a_repo.url");
    expect(scrollMock).toHaveBeenCalledTimes(1);
  });
});
