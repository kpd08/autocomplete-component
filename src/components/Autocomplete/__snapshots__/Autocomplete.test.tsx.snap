// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should render as expected 1`] = `
<div>
  <div
    class="flex flex-col items-center"
  >
    <input
      class="border-2 rounded p-4 max-w-md w-full outline-none"
      placeholder="Search usernames or repositories"
      value=""
    />
    <div
      class="flex flex-col max-w-md w-full max-h-autocompleteResults overflow-auto shadow-md rounded-b-md"
    />
  </div>
</div>
`;
